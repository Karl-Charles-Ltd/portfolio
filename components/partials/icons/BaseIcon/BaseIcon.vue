<template>
  <component :is="iconComponent" class="v-icon" :class="iconClasses" v-bind="$attrs" />
</template>

<script lang="ts">
import { Prop, Component, mixins } from 'nuxt-property-decorator';
import iconMixin from '../iconMixin.vue';
import * as icons from '../icons';

const availableSizes: string[] = ['mini', 'teeny', 'tiny', 'small', 'base', 'medium', 'intermediate', 'big'];

@Component({
  components: { ...icons },
})
export default class BaseIcon extends mixins(iconMixin) {
  @Prop({ required: true, validator: (value) => Object.keys(icons).includes(value) }) readonly icon!: string;
  @Prop({ default: 'teeny', validator: (value) => availableSizes.includes(value) })
  readonly size!: string;

  get iconClasses() {
    return {
      [`v-icon--${this.direction}`]: this.direction && this.icon === 'ArrowCenter',
      [`v-icon--${this.size}`]: this.size,
      [`u-color-${this.theme}`]: this.theme,
    };
  }

  get iconComponent() {
    if (icons[this.icon]) {
      return icons[this.icon];
    } else {
      throw new Error(`${this.icon} does not exist.`);
    }
  }
}
</script>

<style lang="scss">
@import '../setup.scss';

.v-icon {
  fill: currentColor;
  vertical-align: top;

  &--left {
    transform: rotate(90deg);
  }

  &--right {
    transform: rotate(270deg);
  }

  &--up {
    transform: rotate(180deg);
  }

  @each $_size-name, $_size-value in $icon-size-map {
    &.v-icon--#{$_size-name} {
      height: #{$_size-value};
      width: #{$_size-value};
    }
  }
}
</style>
