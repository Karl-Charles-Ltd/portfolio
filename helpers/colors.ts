/**
 * Get the hex color codes that are used throughout the project.
 * These colors and names should match with
 * assets/scss/settings/_settings.colors.scss.
 *
 * @param {String} colorName - The color name from the theme
 * @returns {String} - The hex code for the matching color
 */
export function colorFromTheme(colorName) {
  const colors = {
    primary: '#6b2f93',
    secondary: '#D0438F',
    tertiary: '#09ECFD',
    quaternary: '#f40ee3',
    quinary: '#370b3c',
    black: '#000000',
    charcoal: '#4D4D4D',
    'grey-one': '#949494',
    'grey-two': '#CDCDCD',
    'light-grey': '#E2E2E2',
    'light-grey-one': '#E2E2E2',
    'light-grey-two': '#F9F9F9',
    white: '#FFFFFF',
  };

  return colors[colorName];
}

export const themeMap = {
  primary: {
    background: 'primary',
    foreground: 'secondary',
  },
  'primary-dark': {
    background: 'primary-dark',
    foreground: 'secondary',
  },
  secondary: {
    background: 'secondary',
    foreground: 'white',
    accent: 'tertiary',
  },
  'secondary-dark': {
    background: 'secondary-dark',
    foreground: 'white',
  },
  'secondary-light': {
    background: 'secondary-light',
    foreground: 'white',
  },
  tertiary: {
    background: 'tertiary',
    foreground: 'secondary',
  },
  quaternary: {
    background: 'quaternary',
    foreground: 'secondary',
  },
  quinary: {
    background: 'quinary',
    foreground: 'white',
  },
  'quinary-darker': {
    background: 'quinary-darker',
    foreground: 'white',
  },
  white: {
    background: 'white',
    foreground: 'secondary',
  },
  charcoal: {
    background: 'charcoal',
    foreground: 'white',
  },
  'light-grey-one': {
    background: 'light-grey-one',
    foreground: 'secondary',
  },
};
