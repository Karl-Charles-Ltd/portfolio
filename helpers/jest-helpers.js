import { shallowMount, createLocalVue } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer';
import VueRouter from 'vue-router';

export function setup() {
  const renderer = createRenderer();
  const localVue = createLocalVue();
  const router = new VueRouter();

  // setupPlugins(useVuex, localVue);

  return {
    renderer,
    router,
    localVue,
    shallowMount,
  };
}

export function componentFactory(component, options) {
  return shallowMount(component, options);
}

export function createSnapshot(component, options) {
  const wrapper = componentFactory(component, options);
  expect(wrapper.exists()).toBeTruthy();
  expect(wrapper.html()).toMatchSnapshot();
}
