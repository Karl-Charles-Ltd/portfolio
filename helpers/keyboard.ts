export function addKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}

export function removeKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.removeEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}
