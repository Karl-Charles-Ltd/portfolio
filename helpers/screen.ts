import * as screenBreakpoints from '~/constants/screen-breakpoints';

export const breakpoints: string[] = [
  screenBreakpoints.MOBILE,
  screenBreakpoints.TABLET,
  screenBreakpoints.SMALL_DESK,
  screenBreakpoints.LARGE_DESK,
  screenBreakpoints.HUGE_DESK,
];

interface Screen {
  [key: string]: string | null;
}

export default function mq({ from = null, until = null }: Screen, currentScreenSize: string): boolean {
  const currentBreakpointIndex = breakpoints.indexOf(currentScreenSize);

  if (from !== '' && from !== null) {
    const fromBreakpointIndex = breakpoints.indexOf(from || '');
    if (fromBreakpointIndex > currentBreakpointIndex) {
      return false;
    }
  }

  if (until !== '' && until !== null) {
    const untilBreakpointIndex = breakpoints.indexOf(until || '');
    if (untilBreakpointIndex <= currentBreakpointIndex) {
      return false;
    }
  }

  return true;
}
