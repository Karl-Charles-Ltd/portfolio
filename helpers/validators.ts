import { themeMap } from './colors';

export function validateTheme(theme: string, extraColors: string[] = []): boolean {
  const baseThemes: string[] = Object.keys(themeMap);
  const availableThemes: string[] = baseThemes.concat(extraColors);

  return availableThemes.includes(theme);
}
