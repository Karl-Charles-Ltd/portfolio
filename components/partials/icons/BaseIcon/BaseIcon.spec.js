import BaseIcon from './BaseIcon.vue';
import { setup, componentFactory, createSnapshot, testComputedProperty } from '~/helpers/jest-helpers';

const { localVue } = setup();

const shallowMountComponent = (options) => componentFactory(BaseIcon, { localVue, ...options });

describe('BaseIcon', () => {
  test('is a Vue instance', () => {
    createSnapshot(BaseIcon, { propsData: { icon: 'Bin' } });
  });

  describe('computed', () => {
    describe('iconClasses property', () => {
      let tempVue;
      const expectedIconClasses = {
        direction: 'v-icon--down',
        size: 'v-icon--intermediate',
        theme: 'u-color-quinary',
      };
      beforeEach(() => {
        tempVue = {
          direction: 'down',
          icon: 'ArrowCenter',
          size: 'intermediate',
          theme: 'quinary',
        };
      });
      it('returns the correct direction class if icon is ArrowCenter', () => {
        const iconClasses = testComputedProperty(BaseIcon, 'iconClasses', tempVue);
        expect(iconClasses[expectedIconClasses.direction]).toBeTruthy();
      });
      it('returns correct icon size', () => {
        const iconClasses = testComputedProperty(BaseIcon, 'iconClasses', tempVue);
        expect(iconClasses[expectedIconClasses.size]).toBeTruthy();
      });
      it('returns correct icon theme', () => {
        const iconClasses = testComputedProperty(BaseIcon, 'iconClasses', tempVue);
        expect(iconClasses[expectedIconClasses.theme]).toBeTruthy();
      });
    });
  });

  test('throws an error if the icon does not exist', () => {
    /* eslint-disable no-console */
    const originalError = console.error;
    console.error = jest.fn();

    expect(() => {
      shallowMountComponent({ propsData: { icon: 'Test' } });
    }).toThrowError('Test does not exist.');

    console.error = originalError;
    /* eslint-enable no-console */
  });
});
