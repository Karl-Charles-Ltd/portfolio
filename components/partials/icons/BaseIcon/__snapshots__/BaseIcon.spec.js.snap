// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseIcon is a Vue instance 1`] = `"<anonymous-stub class=\\"v-icon v-icon--teeny u-color-primary\\"></anonymous-stub>"`;
